File: src/main/java/com/partymakers/shareparty/data/persistence/helper/RepositoryFacade.java
package com.partymakers.shareparty.data.persistence.helper;

import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.repository.CrudRepository;

import java.util.Optional;
import java.util.stream.Collectors;

public class RepositoryFacade <T, U, R> implements CrudRepository<T, R> {

    private JpaRepository<U, R>     repository;
    private DomainMapper<T, U>      domainMapper;
    private PersistenceMapper<U, T> persistenceMapper;

    public RepositoryFacade(JpaRepository<U, R> repository, DomainMapper<T, U> domainMapper, PersistenceMapper<U, T> persistenceMapper) {
        this.repository = repository;
        this.domainMapper = domainMapper;
        this.persistenceMapper = persistenceMapper;
    }

    @Override
    public <S extends T> S save(S entity) {
        return (S) persistenceMapper.toDomain(repository.save(domainMapper.toPersistence(entity)));
    }

    @Override
    public <S extends T> Iterable<S> saveAll(Iterable<S> entities) {

        entities.forEach(entity -> save(entity));
        return entities;
    }

    @Override
    public Optional<T> findById(R r) {
        return repository.findById(r).map((persistence) -> persistenceMapper.toDomain(persistence));
    }

    @Override
    public boolean existsById(R r) {
        return repository.existsById(r);
    }

    @Override
    public Iterable<T> findAll() {
         return repository.findAll()
            .stream()
            .map( (persistence) -> persistenceMapper.toDomain(persistence))
            .collect(Collectors.toList());
    }

    @Override
    public Iterable<T> findAllById(Iterable<R> rs) {
        return repository.findAllById(rs)
            .stream()
            .map((persistence) -> persistenceMapper.toDomain(persistence))
            .collect(Collectors.toList());
    }

    @Override
    public long count() {
        return repository.count();
    }

    @Override
    public void deleteById(R r) {
        repository.deleteById(r);
    }

    @Override
    public void delete(T entity) {
        repository.delete(domainMapper.toPersistence(entity));
    }

    @Override
    public void deleteAllById(Iterable<? extends R> entities) {
        repository.deleteAllById(entities);
    }

    @Override
    public void deleteAll(Iterable<? extends T> entities) {
        entities.forEach(entity -> delete(entity));
    }

    @Override
    public void deleteAll() {
        repository.deleteAll();
    }
}

File: src/main/kotlin/com/partymakers/shareparty/friends/data/repository/RepositoryFacade.kt
package com.partymakers.shareparty.friends.data.repository

import com.partymakers.shareparty.friends.data.mapper.DomainMapper
import com.partymakers.shareparty.friends.data.mapper.PersistenceMapper
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.repository.CrudRepository
import java.util.Optional

class RepositoryFacade<T, U, R>(
    private val repository: JpaRepository<U, R>,
    private val domainMapper: DomainMapper<T, U>,
    private val persistenceMapper: PersistenceMapper<U, T>
) : CrudRepository<T, R> {

    override fun <S : T> save(entity: S): S =
        persistenceMapper.toDomain(repository.save(domainMapper.toPersistence(entity))) as S

    override fun <S : T> saveAll(entities: MutableIterable<S>): MutableIterable<S> {
        entities.forEach { save(it) }
        return entities
    }

    override fun findById(id: R): Optional<T> =
        repository.findById(id).map(persistenceMapper::toDomain)

    override fun existsById(id: R): Boolean =
        repository.existsById(id)

    override fun findAll(): MutableIterable<T> =
        repository.findAll()
            .map(persistenceMapper::toDomain)
            .toMutableList()

    override fun findAllById(ids: MutableIterable<R>): MutableIterable<T> =
        repository.findAllById(ids)
            .map(persistenceMapper::toDomain)
            .toMutableList()

    override fun count(): Long =
        repository.count()

    override fun deleteById(id: R) =
        repository.deleteById(id)

    override fun delete(entity: T) =
        repository.delete(domainMapper.toPersistence(entity))

    override fun deleteAllById(ids: MutableIterable<R>) =
        repository.deleteAllById(ids)

    override fun deleteAll(entities: MutableIterable<T>) {
        entities.forEach { delete(it) }
    }

    override fun deleteAll() =
        repository.deleteAll()
}